x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java

```diff
@@ -155,6 +177,89 @@ static EqlSearchResponse createResponse(Results results, AsyncExecutionId id) {
+        return request.waitForCompletionTimeout() != null && request.waitForCompletionTimeout().getMillis() >= 0;
+    }
+
+    // can the request be proxied to the remote cluster?
+    private static boolean canMinimizeRountrips(EqlSearchRequest request, Set<String> clusterAliases) {
+        // Has minimizing the round trips been (explicitly) disabled?
+        if (request.ccsMinimizeRoundtrips() == false) {
+            return false;
+        }
+        // Is this a search against a single, remote cluster?
+        if (clusterAliases.size() != 1 || clusterAliases.contains(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY)) {
+            return false;
+        }
+        // The remote async ID would not be valid on local cluster; furthermore on results fetching we would know neither if the ID is
+        // remote or not, nor which remote cluster it belongs to (TODO: could rewrite the ID to smth like [alias:ID])
+        return requestIsAsync(request) == false;
+    }
+
+    // fixes the _index values by prefixing them with the source cluster alias' name
+    private static EqlSearchResponse qualifyHits(EqlSearchResponse r, String clusterAlias) {
+        EqlSearchResponse.Hits hits = r.hits();
+        List<EqlSearchResponse.Sequence> sequences = qualifySequences(hits.sequences(), clusterAlias);
+        List<EqlSearchResponse.Event> events = sequences != null ? null : qualifyEvents(hits.events(), clusterAlias, true);
+        return new EqlSearchResponse(new EqlSearchResponse.Hits(events, sequences, hits.totalHits()), r.took(), r.isTimeout());
+    }
+
+    private static List<EqlSearchResponse.Sequence> qualifySequences(List<EqlSearchResponse.Sequence> sequences, String clusterAlias) {
+        List<EqlSearchResponse.Sequence> qualifiedSequences = null;
+        if (sequences != null && sequences.isEmpty() == false) {
+            qualifiedSequences = new ArrayList<>(sequences.size());
+            for (EqlSearchResponse.Sequence s : sequences) {
+                qualifiedSequences.add(new EqlSearchResponse.Sequence(s.joinKeys(), qualifyEvents(s.events(), clusterAlias, false)));
+            }
+        }
+        return qualifiedSequences;
+    }
+
+    private static List<EqlSearchResponse.Event> qualifyEvents(List<EqlSearchResponse.Event> events, String clusterAlias,
+                                                               boolean emptyListIfNoEvents) {
+        List<EqlSearchResponse.Event> qualifiedEvents = null;
+        if (events != null && events.isEmpty() == false) {
+            qualifiedEvents = new ArrayList<>(events.size());
+            for (EqlSearchResponse.Event e : events) {
+                qualifiedEvents.add(new EqlSearchResponse.Event(buildRemoteIndexName(clusterAlias, e.index()), e.id(), e.source(),
+                    e.fetchFields()));
+            }
+        }
+        return qualifiedEvents == null && emptyListIfNoEvents ? new ArrayList<>() : qualifiedEvents;
+    }
```

Comment:
Unless I'm missing something, these methods copy the response to prefix the index which is a waste.
Either change the actual returned objects to be mutable so that the remote cluster alias is added to the index or push the clusterAlias info so it's available during the `EqlSearchResponse.*` creation.

Reply:
I've change the mutability to allow setting an index name. The alternative would require passing the cluster name thorough multiple layer, which is doable, but seemed a bit more contrived.